### PR TITLE
feat: support configuring the order of the upstream dns service discovery resolve

### PR DESCRIPTION
**Pull request by monkeyDluffy6017**
…very resolve

### Description

Some users are having problems with the upstream DNS service discovery feature, because the current DNS resolution order is fixed:  1. "last" 2. "SRV" 3. "A" 4. "AAAA"  5. "CNAME".
"last" means the last successful resolving method; that is, if the last resolving is successful, the last resolving method will be used first.
This causes the problem that if for some reason the A record is resolved successfully, the higher priority SRV record will not be resolved in the future, and the user expects to get the result of resolving the SRV record.

As a user, I want to configure how and in what order the upstream DNS service discovery function resolves, and APISIX is able to resolve in the way and in the order I configure.

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Review by tokers — COMMENTED** [review]
@monkeyDluffy6017 We should also update the https://apisix.apache.org/docs/apisix/discovery/dns/#service-discovery-via-dns.

File: apisix/discovery/dns/schema.lua [hunk 1]
```diff
@@ -24,6 +24,18 @@ return {
...
             },
         },
+        order = {
+            type = "array",
+            minItems = 1,
+            maxItems = 5,
+            uniqueItems = true,
+            items = {
+                enum = {"last", "SRV", "A", "AAAA", "CNAME"},
+     --           ["not"] = {
```
Comment (spacewander) [inline apisix/discovery/dns/schema.lua:34]: Just for debug?

File: t/discovery/dns/sanity.t [hunk 2]
```diff
@@ -318,3 +318,24 @@ qr/upstream nodes: \{[^}]+\}/
...
+    - service_name: "srv-a.test.local"
+      discovery_type: dns
+      type: roundrobin
+      id: 1
+discovery:
+    dns:
+        servers:
+            - "127.0.0.1:1053"
+        order:
+            - A
```
Comment (spacewander) [inline t/discovery/dns/sanity.t:336]: Is it possible to add some test about invalid input, like `a` or repeated record type?

File: apisix/discovery/dns/schema.lua (hunk 1 above)
Comment (monkeyDluffy6017) [reply apisix/discovery/dns/schema.lua]: Done

File: t/discovery/dns/sanity.t (hunk 2 above)
Comment (monkeyDluffy6017) [reply t/discovery/dns/sanity.t]: Done

File: t/discovery/dns/sanity.t [hunk 3]
```diff
@@ -318,3 +326,115 @@ qr/upstream nodes: \{[^}]+\}/
...
+            - SRV
+--- apisix_yaml
+upstreams:
+    - service_name: "srv-a.test.local"
+      discovery_type: dns
+      type: roundrobin
+      id: 1
+--- must_die
+--- error_log
+matches none of the enum values
```
Comment (tzssangglass) [inline t/discovery/dns/sanity.t:440]: Can we add a test case: DNS Server returns SRV records first for the same domain name, and the next query returns A records?

**Comment by monkeyDluffy6017** [comment]
> 

Done

**Review by spacewander — APPROVED** [review]

**Review by membphis — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**